File: src/main/java/com/koddy/server/member/domain/model/mentor/UniversityAuthentication.java
package com.koddy.server.member.domain.model.mentor;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import jakarta.persistence.Enumerated;

import static com.koddy.server.member.domain.model.mentor.AuthenticationStatus.ATTEMPT;
import static com.koddy.server.member.domain.model.mentor.AuthenticationStatus.SUCCESS;
import static jakarta.persistence.EnumType.STRING;

@Embeddable
public class UniversityAuthentication {
    protected UniversityAuthentication() {
    }

    @Column(name = "school_mail")
    private String schoolMail;

    @Column(name = "proof_data_upload_url")
    private String proofDataUploadUrl;

    @Enumerated(STRING)
    @Column(name = "auth_status", columnDefinition = "VARCHAR(20)")
    private AuthenticationStatus status;

    private UniversityAuthentication(
            final String schoolMail,
            final String proofDataUploadUrl,
            final AuthenticationStatus status
    ) {
        this.schoolMail = schoolMail;
        this.proofDataUploadUrl = proofDataUploadUrl;
        this.status = status;
    }

    public static UniversityAuthentication attemptMail(final String schoolMail) {
        return new UniversityAuthentication(schoolMail, null, ATTEMPT);
    }

    public static UniversityAuthentication attemptProofData(final String proofDataUploadUrl) {
        return new UniversityAuthentication(null, proofDataUploadUrl, ATTEMPT);
    }

    public void complete() {
        status = SUCCESS;
    }

    public boolean isAuthenticated() {
        return status == SUCCESS;
    }

    public String getSchoolMail() {
        return schoolMail;
    }

    public String getProofDataUploadUrl() {
        return proofDataUploadUrl;
    }

    public AuthenticationStatus getStatus() {
        return status;
    }
}

File: src/main/kotlin/com/koddy/server/member/domain/model/mentor/UniversityAuthentication.kt
package com.koddy.server.member.domain.model.mentor

import com.koddy.server.member.domain.model.mentor.UniversityAuthentication.AuthenticationStatus.ATTEMPT
import com.koddy.server.member.domain.model.mentor.UniversityAuthentication.AuthenticationStatus.SUCCESS
import jakarta.persistence.Column
import jakarta.persistence.Embeddable
import jakarta.persistence.EnumType.STRING
import jakarta.persistence.Enumerated

@Embeddable
data class UniversityAuthentication(
    @Column(name = "school_mail")
    val schoolMail: String?,

    @Column(name = "proof_data_upload_url")
    val proofDataUploadUrl: String?,

    @Enumerated(STRING)
    @Column(name = "auth_status", columnDefinition = "VARCHAR(20)")
    val status: AuthenticationStatus,
) {
    val isAuthenticated: Boolean
        get() = status == SUCCESS

    fun complete(): UniversityAuthentication {
        return UniversityAuthentication(schoolMail, proofDataUploadUrl, SUCCESS)
    }

    companion object {
        fun attemptMail(schoolMail: String): UniversityAuthentication {
            return UniversityAuthentication(
                schoolMail = schoolMail,
                proofDataUploadUrl = null,
                status = ATTEMPT,
            )
        }

        fun attemptProofData(proofDataUploadUrl: String): UniversityAuthentication {
            return UniversityAuthentication(
                schoolMail = null,
                proofDataUploadUrl = proofDataUploadUrl,
                status = ATTEMPT,
            )
        }
    }

    enum class AuthenticationStatus {
        ATTEMPT,
        SUCCESS,
    }
}
